File: 프로그래머스/lv3/49191. 순위/순위.java
import java.util.*;

class Solution {
    private final int MAX = 10000;
    
    public int solution(int n, int[][] results) {
        int[][] graph = new int[n+1][n+1];
        
        for (int i=0; i<graph.length; i++) {
            for (int j=0; j<graph[i].length; j++) {
                if (i != j) {
                    graph[i][j] = MAX;
                }
            }
        }
        
        for (int[] result : results) {
            graph[result[0]][result[1]] = 1;
        }
        
        for (int k=1; k<graph.length; k++) {
            for (int i=1; i<graph.length; i++) {
                for (int j=1; j<graph[i].length; j++) {
                    graph[i][j] = Math.min(graph[i][j], graph[i][k] + graph[k][j]);
                }
            }
        }
        
        int answer = n;
        for (int i=1; i<graph.length; i++) {
            for (int j=1; j<graph[i].length; j++) {
                if (i == j) continue;
                
                if (graph[i][j] == MAX && graph[j][i] == MAX) {
                    answer--;
                    break;
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/lv3/49191. 순위/순위.kt
class Solution {
    fun solution(n: Int, results: Array<IntArray>): Int {
        val INF = Int.MAX_VALUE
        
        // 그래프 초기화 (노드간 거리는 INF, 자기 자신은 0)
        val graph = Array(n + 1) { IntArray(n + 1) { if (it == 0) 0 else INF } }
        
        // 주어진 edges 정보로 그래프에 경로 설정
        for (result in results) {
            graph[result[0]][result[1]] = 1
        }
        
        floydWarshall(n, graph)
        
        return (1..n).count { node ->
            val count = (1..n).count { other ->
                graph[node][other] != INF || graph[other][node] != INF
            }
            count == n - 1
        }
    }
    
    fun floydWarshall(nodeCount: Int, graph: Array<IntArray>) {
        for (k in 1..nodeCount) {
            for (i in 1..nodeCount) {
                for (j in 1..nodeCount) {
                    if (graph[i][k] != Int.MAX_VALUE && graph[k][j] != Int.MAX_VALUE) {
                        graph[i][j] = minOf(graph[i][j], graph[i][k] + graph[k][j])
                    }
                }
            }
        }
    }
}